### DISCUSSION TITLE
Files Service: connection aborted during download of specific files

### DISCUSSION THREAD

**Discussion by edocrippaofficial** [mia-platform/community#304] [Q&A] [answered]
### Which section?

Marketplace

### Which is the version?

2.8.0

### Description

Hi, recently in production environment we found some errors downloading specific files from files-service (based on gridfs).
Basically the request for the download results in an aborted connection.

Here are some cURL request that I did from another pod in the same namespace:
```
curl -v http://files-service/download/64f2ef5028f93e6c5c394c4f.pdf?downloadWithOriginalName=1 > file.pdf
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 10.4.15.182:80...
* Connected to files-service (10.4.15.182) port 80 (#0)
> GET /download/64f2ef5028f93e6c5c394c4f.pdf?downloadWithOriginalName=1 HTTP/1.1
> Host: files-service
> User-Agent: curl/7.88.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< content-type: application/pdf
< content-disposition: attachment; filename="doc.pdf"
< x-accel-buffering: no
< Date: Mon, 04 Sep 2023 07:36:27 GMT
< Connection: keep-alive
< Keep-Alive: timeout=72
< Transfer-Encoding: chunked
< 
{ [13821 bytes data]
* transfer closed with outstanding read data remaining
100  219k    0  219k    0     0  8370k      0 --:--:-- --:--:-- --:--:-- 8451k
* Closing connection 0
curl: (18) transfer closed with outstanding read data remaining
```

As you can see I got the cURL error 18, meaning that the connection was closed without sending all the chunks correctly.

The strange thing is that the number of bytes sent is not always the same, as in this other example:
```
* transfer closed with outstanding read data remaining
100  199k    0  199k    0     0  8405k      0 --:--:-- --:--:-- --:--:-- 8657k
* Closing connection 0
curl: (18) transfer closed with outstanding read data remaining
```

From a request made by Node.js the stream goes in the `aborted` state.  
It happens once every hundreds of file, but with that file the outcome is the same: connection aborted.

Another very strange thing is that if the request is made to the pod via port-forward, with the same file, the response is correct:
```
curl -v http://localhost:3000/download/64f2ef5028f93e6c5c394c4f.pdf?downloadWithOriginalName=1 > file.pdf
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1:3000...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 3000 (#0)
> GET /download/64f2ef5028f93e6c5c394c4f.pdf?downloadWithOriginalName=1 HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: application/pdf
< content-disposition: attachment; filename="doc.pdf"
< x-accel-buffering: no
< Date: Mon, 04 Sep 2023 07:50:57 GMT
< Connection: keep-alive
< Keep-Alive: timeout=72
< Transfer-Encoding: chunked
< 
{ [32509 bytes data]
100  257k    0  257k    0     0   975k      0 --:--:-- --:--:-- --:--:--  975k
* Connection #0 to host localhost left intact
```

As you can see, I got all the 257k bytes of the file...

I can't investigate further if it's a problem with GridFs, with the connection to Mongo, a problem with the files-service or a network problem inside the cluster.

Any ideas?

### Actual Outcome

The connection goes in `aborted` state

### Expected Outcome

I can download the file

**Comment by theliuk** [comment]
Yes, I only add a couple of notes:
1. the Node js pod and the `files-service` pod resides in the same namespace
2. the connection seems to be closed by the `files-service`
3. the namespace dashboard doesn't show evidence of any dropped packet
4. the mongodb cluster seems to be healthy

**Comment by fredmaggiowski** [comment]
A few questions that come to my mind to further investigate the issue:

 * does this happen only with specific files or it can happen with any file?
 * what API Gateway are you using? (seeing the `x-accel-buffering:` in the response headers I assume it's `nginx`)
 * is there some advanced configuration on the API Gateway
 * does it appear to be "time based"? e.g. the connection always gets dropped after 10s?


@FabioNappi this seems like a similar problem to one your team faced a few weeks ago, am I right?

**Reply by edocrippaofficial** [reply]
Hi @fredmaggiowski 
In oder:
- it happens only with specific files. If a file cannot be downloaded, it remains in that state. Moreover, @theliuk and I once restarted the files-service, and suddenly the same file started to be downloaded successfully 
- the header x-accel-buffering is wrote by files-service, but that has no influence whatsoever in this case, because we're not passing through an api gateway: we are simply making an http request from a pod to another
- as above, the api gateway is not involved
- no, it's not ime based, as it drops after like 10ms

**Reply by FabioNappi** [reply]
Hi @fredmaggiowski , @EdoCrippaOfficial 
yes we encountered the exact same issue using Mongo GridFS. To solve this problem we had to change the bucket provider: switching to Google Cloud Storage we no longer faced the issue.

**Reply by theliuk** [reply]
We'll add some logs to the service, hoping to get some insights about the cause of the error.

**Comment by theliuk** [comment]
The team lead by @pmarchini may have found a solution to [the issue](https://makeitapp.atlassian.net/browse/CPR-151).
We tried a beta version of the files-service image with the fix and it works.

**Reply by theliuk** [reply, marked as answer]
To fix the issue upgrade files-service to version `^2.9.3`.
